src/qililab/instruments/qdevil/qdevil_qdac2.py

```diff
@@ -102,6 +105,58 @@ def set_parameter(self, parameter: Parameter, value: ParameterValue, channel_id:
             return
         raise ParameterNotFound(self, parameter)
 
+    def get_dac(self, channel_id: ChannelID):
+        """Get specific DAC from QDAC.
+
+        Args:
+            channel_id (ChannelID): channel id of the dac
+        """
+        return self.device.channel(channel_id)
+
+    def upload_waveform(self, waveform: Waveform, channel_id: ChannelID):
+        """Uploads a waveform to the instrument and saves it to _cache.
+        IMPORTANT: note that the waveform resolution is not to the ns, it is acutally around 1_micro_second.
+
+        Args:
+            waveform (Waveform): Waveform to upload
+            channel_id (ChannelID): channel id of the qdac
+
+        Raises:
+            ValueError: if a waveform is already allocated
+        """
+        values = list(waveform.envelope())  # TODO: does np array work?
+        if channel_id in self._cache:
+            raise ValueError(
+                f"Device {self.name} already has a waveform allocated to channel {channel_id}. Clear the cache before allocating a new waveform"
+            )
+        trace = self.device.allocate_trace(channel_id, len(values))
```

Comment:
This is a change we need now, traces length must be even values. We need to check for that or otherwise raise an warning for the moment, but maybe an error is better as it does duplicate the last point in the current firmware but they might change that in future firmware updates:

"The maximum length of a single trace is 6
million points (6,291,456). The length of traces should be an even number.
Note that with the current firmware version (13-1.57), no error is produced when uploading traces with
uneven number of points. Instead, the last point is duplicated."

Also we need to check the range is between [-1, 1]:
"Trace values should be in the interval [-1;1]. Their scaling into actual volts is handled by the AWG
generator of the relevant channels. "

Reply:
thanks for the fix!